### PR TITLE
[SPARK-12288] [SQL] Support UnsafeRow in Coalesce/Except/Intersect.

### PR DESCRIPTION
**Pull request by gatorsmile**
Support UnsafeRow for the Coalesce/Except/Intersect.

Could you review if my code changes are ok? @davies Thank you!

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala [hunk 1]
```diff
@@ -263,6 +265,10 @@ case class Except(left: SparkPlan, right: SparkPlan) extends BinaryNode {
   protected override def doExecute(): RDD[InternalRow] = {
     left.execute().map(_.copy()).subtract(right.execute().map(_.copy()))
   }
+
+  override def outputsUnsafeRows: Boolean = children.forall(_.outputsUnsafeRows)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala]: If an operator's children produce both unsafe and safe rows, we will convert everything unsafe rows.
so here should be `children.exists(_.outputsUnsafeRows)`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala [hunk 2]
```diff
@@ -263,6 +265,10 @@ case class Except(left: SparkPlan, right: SparkPlan) extends BinaryNode {
...
+  override def canProcessUnsafeRows: Boolean = true
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala]: UnsafeRow can be compared with GenericInternalRow, so we should make sure that `left` and `right` have the same type of row.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/RowFormatConvertersSuite.scala [hunk 3]
```diff
@@ -58,6 +58,48 @@ class RowFormatConvertersSuite extends SparkPlanTest with SharedSQLContext {
...
   }
 
+  test("coalesce can process unsafe rows") {
+    val plan = Coalesce(1, outputsUnsafe)
+    val preparedPlan = sqlContext.prepareForExecution.execute(plan)
+    assert(getConverters(preparedPlan).size === 1)
+    assert(preparedPlan.outputsUnsafeRows)
+  }
+
+  test("coalesce can process safe rows") {
```
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/RowFormatConvertersSuite.scala]: We may remove the safe rows in future, it's fine to not have these tests (for safe rows).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala]: seems it's already promised in https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/rowFormatConverters.scala#L95-L106

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala]: Thank you! I also found the same issue exists in ```Union``` I also did a correction.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala]: Thank you! @davies @cloud-fan I added two extra test cases for ensuring they have the same formats.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/RowFormatConvertersSuite.scala (hunk 3 above)
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/RowFormatConvertersSuite.scala]: uh,,, I see! : )

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
The failed test case is not related to my change. I will try to see if I can fix the flaky test. 

retest this please

**Comment by gatorsmile** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/RowFormatConvertersSuite.scala (hunk 3 above)
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/RowFormatConvertersSuite.scala]: Will you remove these?

Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/RowFormatConvertersSuite.scala]: Sorry, doing it now.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/RowFormatConvertersSuite.scala [hunk 4]
```diff
@@ -58,6 +58,41 @@ class RowFormatConvertersSuite extends SparkPlanTest with SharedSQLContext {
     assert(!preparedPlan.outputsUnsafeRows)
   }
 
+  test("coalesce can process unsafe rows") {
+    val plan = Coalesce(1, outputsUnsafe)
+    val preparedPlan = sqlContext.prepareForExecution.execute(plan)
+    assert(getConverters(preparedPlan).size === 1)
+    assert(preparedPlan.outputsUnsafeRows)
+  }
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/RowFormatConvertersSuite.scala]: should we add mixed case test for coalesce too?

Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/RowFormatConvertersSuite.scala]: Sorry, I am not sure if I understand your question. `Coalesce` only have a single child. I am not sure what mixed rows mean. Could you give me one example? Thank you!

Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/RowFormatConvertersSuite.scala]: oh sorry, I thought it was `org.apache.spark.sql.catalyst.expressions.Coalesce`, nvm

Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/RowFormatConvertersSuite.scala]: Thank you!

**Comment by davies** [comment]
LGTM, merging into master, thanks!

**Pull request closed without merging**